### PR TITLE
EnC: Fix document out-of-sync checks when module is not loaded when d…

### PR DESCRIPTION
**Pull request by tmat**
…ebugging session starts

Previous change https://github.com/dotnet/roslyn/pull/39295 allowed us to rely on PDB source file checksums when distinguishing between design-time-only documents that should be ignored when applying changes and documents whose changes must be applied to the debuggee. The change did not handle the case when the PDB isn't loaded yet when this check is being performed. This bug resulted in changes not being applied correctly and documents getting out-of-sync in scenarios where the modules were not loaded to the debuggee fast enough.

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1013242

### CODE REVIEW

**Comment by tmat** [comment]
@ivanbasov @jasonmalinowski PTAL

File: src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs [hunk 1]
```diff
@@ -32,9 +32,32 @@ internal sealed class CommittedSolution
         internal enum DocumentState
         {
             None = 0,
-            OutOfSync = 1,
-            MatchesDebuggee = 2,
-            DesignTimeOnly = 3,
+
+            /// <summary>
+            /// The document belongs to a project whose compiled module has not been loaded yet.
+            /// This document state may change to to <see cref="OutOfSync"/>, <see cref="MatchesDebuggee"/>
```
Comment (ivanbasov) [inline src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs]: >to to [](start = 47, length = 5)

double to

**Comment by tmat** [comment]
Although this change fixes the linked bug, there are some more cases I'm working thru...

File: src/Features/Core/Portable/Debugging/DebugInformationReaderProvider.cs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Immutable;
```
Comment (ivanbasov) [inline src/Features/Core/Portable/Debugging/DebugInformationReaderProvider.cs]: need?

File: src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs [hunk 3]
```diff
@@ -32,9 +32,42 @@ internal sealed class CommittedSolution
...
-            OutOfSync = 1,
-            MatchesDebuggee = 2,
-            DesignTimeOnly = 3,
+
+            /// <summary>
+            /// The current document content matches the content the built module was compiled with.
+            /// The document content is matched with the build output instead of the loaded module
+            ///  since the module hasn't been loaded yet.
+            ///
+            /// This document state may change to to <see cref="OutOfSync"/> or <see cref="MatchesDebuggee"/>
```
Comment (ivanbasov) [inline src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs]: >to to [](start = 47, length = 5)

to to

File: src/Test/Utilities/Portable/FX/ImmutableArrayTestExtensions.cs [hunk 4]
```diff
@@ -13,29 +13,33 @@ namespace Microsoft.CodeAnalysis.Test.Utilities
...
 
-                int offset = 0;
-                while (offset < bytes.Length)
-                {
-                    int length = Math.Min(bufferSize, bytes.Length - offset);
-                    bytes.CopyTo(offset, buffer, 0, length);
-                    fileStream.Write(buffer, 0, length);
-                    offset += length;
-                }
+            // PERF: Consider using an ObjectPool<byte[]> here
```
Comment (ivanbasov) [inline src/Test/Utilities/Portable/FX/ImmutableArrayTestExtensions.cs]: >// PERF: Consider using an ObjectPool<byte[]> here [](start = 12, length = 50)

TODO? Work item?

Comment (tmat) [reply src/Test/Utilities/Portable/FX/ImmutableArrayTestExtensions.cs]: This comment was there before my changes and it's just a test code anyways.

File: src/EditorFeatures/Test/EditAndContinue/EditAndContinueMethodDebugInfoReaderTests.cs [hunk 5]
```diff
@@ -95,6 +96,15 @@ public static void Main()
...
             Assert.Equal(default, localSig);
+
+            // document checksums:
+            Assert.False(reader.TryGetDocumentChecksum("/b/c.cs", out _, out _));
+            Assert.False(reader.TryGetDocumentChecksum("/a/d.cs", out _, out _));
+            Assert.False(reader.TryGetDocumentChecksum("/A/C.cs", out _, out _));
+
+            Assert.True(reader.TryGetDocumentChecksum("/a/c.cs", out var actualChecksum, out var actualAlgorithm));
+            Assert.Equal("21-C8-B2-D7-A3-6B-49-C7-57-DF-67-B8-1F-75-DF-6A-64-FD-59-22", BitConverter.ToString(actualChecksum.ToArray()));
+            Assert.Equal(new Guid("ff1816ec-aa5e-4d10-87f7-6f4963833460"), actualAlgorithm);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/EditAndContinue/EditAndContinueMethodDebugInfoReaderTests.cs]: Is there a constant for that?

File: src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs [hunk 6]
```diff
@@ -32,9 +32,42 @@ internal sealed class CommittedSolution
...
         {
             None = 0,
-            OutOfSync = 1,
-            MatchesDebuggee = 2,
-            DesignTimeOnly = 3,
+
+            /// <summary>
+            /// The current document content matches the content the built module was compiled with.
+            /// The document content is matched with the build output instead of the loaded module
+            ///  since the module hasn't been loaded yet.
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs]:
```suggestion
            //  since the module hasn't been loaded yet.
```

**Pull request merged**